### PR TITLE
Removed Bert dependency from BertGeneration code base.

### PR DESCRIPTION
**Pull request by Threepointone4**
# What does this PR do?

-  Related to #19303 
-  Removed `BertGeneration` dependency from `Bert` code base.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@sgugger

### CODE REVIEW

File: src/transformers/models/bert_generation/modeling_bert_generation.py [hunk 1]
```diff
@@ -41,6 +43,478 @@
 _TOKENIZER_FOR_DOC = "BertGenerationTokenizer"
 
 
+# Copied from transformers.models.bert.modeling_bert.BertEmbeddings
+class BertEmbeddings(nn.Module):
```
Comment (sgugger) [inline src/transformers/models/bert_generation/modeling_bert_generation.py:47]:
```suggestion
class BertGenerationEmbeddings(nn.Module):
```
All prefixes should be adapted to the new model (so here and all other classes below up to `BertEncoder`->`BertGenerationEncoder`. Some of your copied from statements will then need to be completed (I'll give on example below).

File: src/transformers/models/bert_generation/modeling_bert_generation.py [hunk 2]
```diff
@@ -41,6 +43,478 @@
...
+        context_layer = context_layer.view(new_context_layer_shape)
+
+        outputs = (context_layer, attention_probs) if output_attentions else (context_layer,)
+
+        if self.is_decoder:
+            outputs = outputs + (past_key_value,)
+        return outputs
+
+
+# Copied from transformers.models.bert.modeling_bert.BertAttention
```
Comment (sgugger) [inline src/transformers/models/bert_generation/modeling_bert_generation.py:252]:
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertAttention with Bert->BertGeneration
```
This class uses other BertXxx modules which will be renamed `BertGenerationXxx`, so you will need to add this to the copied from statement.

**Review by sgugger — COMMENTED** [review]
Thanks a lot for your PR! All the layer names should have the "model" prefix, e.g. BertGeneration in this case. I've left pointers on how to fix this (to be repeated for all classes).

File: src/transformers/models/bert_generation/modeling_bert_generation.py [hunk 3]
```diff
@@ -41,6 +43,415 @@
...
+
+        return outputs
+
+    def feed_forward_chunk(self, attention_output):
+        intermediate_output = self.intermediate(attention_output)
+        layer_output = self.output(intermediate_output, attention_output)
+        return layer_output
+
+
+# Copied from transformers.models.bert.modeling_bert.BertEncoder
```
Comment (sgugger) [inline src/transformers/models/bert_generation/modeling_bert_generation.py:357]:
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertEncoder with Bert->BertGeneration
```

File: src/transformers/models/bert_generation/modeling_bert_generation.py [hunk 4]
```diff
@@ -41,6 +43,415 @@
...
+                relative_position_scores = torch.einsum("bhld,lrd->bhlr", query_layer, positional_embedding)
+                attention_scores = attention_scores + relative_position_scores
+            elif self.position_embedding_type == "relative_key_query":
+                relative_position_scores_query = torch.einsum("bhld,lrd->bhlr", query_layer, positional_embedding)
+                relative_position_scores_key = torch.einsum("bhrd,lrd->bhlr", key_layer, positional_embedding)
+                attention_scores = attention_scores + relative_position_scores_query + relative_position_scores_key
+
+        attention_scores = attention_scores / math.sqrt(self.attention_head_size)
+        if attention_mask is not None:
+            # Apply the attention mask is (precomputed for all layers in BertModel forward() function)
```
Comment (sgugger) [inline src/transformers/models/bert_generation/modeling_bert_generation.py:162]:
```suggestion
            # Apply the attention mask is (precomputed for all layers in BertGenerationModel forward() function)
```

**Review by sgugger — COMMENTED** [review]
Those two changes should fix the CI red check :-)

File: src/transformers/models/bert_generation/modeling_bert_generation.py (hunk 3 above)
Comment (Threepointone4) [reply src/transformers/models/bert_generation/modeling_bert_generation.py:357]: @sgugger Thanks for the input. I miss understood that the class which we change names only should have `Bert->BertGeneration`.  I will make this change and push. Thanks

File: src/transformers/models/bert_generation/modeling_bert_generation.py (hunk 4 above)
Comment (Threepointone4) [reply src/transformers/models/bert_generation/modeling_bert_generation.py:162]: @sgugger Thanks for this. I was trying to debug this. When i ran the checks the issue line is showed different. 
`- src/transformers/models/bert_generation/modeling_bert_generation.py: copy does not match models.bert.modeling_bert.BertSelfAttention at line 62
`  
Any reason for that ?

Comment (sgugger) [reply src/transformers/models/bert_generation/modeling_bert_generation.py]: The line it shows you is the line the function/class is defined, not the line of the error. To debug, I personally ran `make fix-copies` and looked at the changes.

**Review by sgugger — APPROVED** [review]

**Comment by sgugger** [comment]
Thanks for your contribution!

File: src/transformers/models/bert_generation/modeling_bert_generation.py [hunk 5]
```diff
@@ -41,6 +43,415 @@
...
+        return outputs
+
+    def feed_forward_chunk(self, attention_output):
+        intermediate_output = self.intermediate(attention_output)
+        layer_output = self.output(intermediate_output, attention_output)
+        return layer_output
+
+
+# Copied from transformers.models.bert.modeling_bert.BertEncoder with Bert->BertGeneration
+class BertEncoder(nn.Module):
```
Comment (SD-13) [inline src/transformers/models/bert_generation/modeling_bert_generation.py:358]: Hi @sgugger, I was looking into this PR for some reference and found that the class name `BertEncoder` is not changed to `BertGenerationEncoder`. I think it was done by mistake. PTAL, Thanks,
/cc @Threepointone4

Comment (sgugger) [reply src/transformers/models/bert_generation/modeling_bert_generation.py:358]: Indeed, this should probably be changed to `BertGenerationEncoder` if you want to make a PR.

Comment (SD-13) [reply src/transformers/models/bert_generation/modeling_bert_generation.py:358]: Okay, I will include it in my new PR.

**Pull request merged**